### PR TITLE
feat(client): Extended query to be able to refer to the genesis and the earliest [oldest] available blocks

### PR DESCRIPTION
**Pull request by frol**
This is based on #2990.

This is necessary for RosettaRPC implementation (#2738) and potentially useful for NEAR Indexer.

Testing Plan
============

Added minimal RPC tests.

### CODE REVIEW

File: chain/client/src/view_client.rs [hunk 1]
```diff
@@ -158,6 +157,28 @@ impl ViewClientActor {
...
+                let tail = chain_store_update.tail()?;
+                for height in tail..=head_header_height {
+                    if let Ok(block_hash) = chain_store_update.get_block_hash_by_height(height) {
+                        return Ok(Some(block_hash));
+                    }
+                }
+                Ok(None)
+            }
+        }
+    }
```
Comment (frol) [inline chain/client/src/view_client.rs:180]: This is the core of the PR (the rest is just renaming of `BlockIdOrFinality` to `BlockReference` as the original naming does not fit the extended semantic).

Comment (Kouprin) [reply chain/client/src/view_client.rs:180]: I guess you want the following:
```
                for height in tail + 1..=head_header_height {
                    if let Ok(block_hash) = chain_store_update.get_block_hash_by_height(height) {
                        return Ok(Some(*chain_store_update.get_block_header(block_hash)?.prev_hash()));
                    }
                }
```
The reason is, the Tail is the position of where GC should continue working. The Tail is always equal of greater than the earliest Block in the Chain because in GC we're looking for the Block X (next one to the earliest one) and remove previous of X (the earliest one) - only this way Tries allow to delete data properly.

File: chain/chain/src/store.rs [hunk 2]
```diff
@@ -142,6 +142,20 @@ pub trait ChainStoreAccess {
...
+        if let Ok(block_hash) = self.get_block_hash_by_height(tail) {
+            return Ok(Some(block_hash.clone()));
+        }
+        for height in tail + 1..=head_header_height {
+            if let Ok(block_hash) = self.get_block_hash_by_height(height) {
+                return Ok(Some(self.get_block_header(&block_hash)?.prev_hash().clone()));
+            }
+        }
+        Ok(None)
+    }
```
Comment (frol) [inline chain/chain/src/store.rs:158]: This is the core of the PR (the rest is just renaming of BlockIdOrFinality to BlockReference as the original naming does not fit the extended semantic).

File: chain/chain/src/store.rs [hunk 3]
```diff
@@ -142,6 +142,20 @@ pub trait ChainStoreAccess {
...
     /// Returns hash of the block on the main chain for given height.
     fn get_block_hash_by_height(&mut self, height: BlockHeight) -> Result<CryptoHash, Error>;
+    /// Returns hash of the first available block after genesis.
+    fn get_earliest_block_hash(&mut self) -> Result<Option<CryptoHash>, Error> {
+        let head_header_height = self.head_header()?.height();
+        let tail = self.tail()?;
+        if let Ok(block_hash) = self.get_block_hash_by_height(tail) {
+            return Ok(Some(block_hash.clone()));
+        }
+        for height in tail + 1..=head_header_height {
```
Comment (evgenykuzyakov) [inline chain/chain/src/store.rs:152]: Let's say the node misses some blocks, how expensive is this loop in the worst case scenario long-term?

Comment (frol) [reply chain/chain/src/store.rs:152]: In the normal case, where a node is mostly online and don't miss blocks (we optimize our RPC nodes for this), this should not cause problems.

@Kouprin What do you think?

Comment (bowenwang1996) [reply chain/chain/src/store.rs:152]: if a node misses some blocks it will either sync the blocks or sync to the network through state sync, in which case tail will be reset. The only case where this can get be expensive is when a lot of heights are skipped. However, since doomslug has liveness, this should not be an issue.

Comment (Kouprin) [reply chain/chain/src/store.rs:152]: > Let's say the node misses some blocks

@evgenykuzyakov We are getting Tail of Canonical Chain which is five epochs behind the Head in average. Not sure if your question is relevant.

@frol I agreed with @bowenwang1996, it should be fine in normal circumstances. In case of skipping lots of Blocks in Canonical Chain for a long time we will face way more unpleasant consequences like crashing the node.

File: chain/chain/src/store.rs [hunk 4]
```diff
@@ -142,6 +142,20 @@ pub trait ChainStoreAccess {
     fn get_block_header(&mut self, h: &CryptoHash) -> Result<&BlockHeader, Error>;
     /// Returns hash of the block on the main chain for given height.
     fn get_block_hash_by_height(&mut self, height: BlockHeight) -> Result<CryptoHash, Error>;
+    /// Returns hash of the first available block after genesis.
+    fn get_earliest_block_hash(&mut self) -> Result<Option<CryptoHash>, Error> {
+        let head_header_height = self.head_header()?.height();
+        let tail = self.tail()?;
+        if let Ok(block_hash) = self.get_block_hash_by_height(tail) {
+            return Ok(Some(block_hash.clone()));
+        }
```
Comment (Kouprin) [inline chain/chain/src/store.rs:151]: This case is covered by the loop below. You can remove it safely.

Comment (frol) [reply chain/chain/src/store.rs:151]: It does not cover the case when there is no block after the tail, though it is a rare corner case, maybe even impossible, though in the test suite we have a network with only the Genesis block and nothing else, so without this check it will return None 🤷‍♂️

What do you recommend?

Comment (Kouprin) [reply chain/chain/src/store.rs:151]: Good point. Could you please add a simple comment explaining this like "Case of having the only Block"?
